### PR TITLE
Enhance specification

### PR DESCRIPTION
**Pull request by otaviojava**
# Changes

- Define the entity
- Define the basic fields
- Define the annotations
- Put sample annotation
- Include the annotations for compatibility with Jakarta Data
- Remove the specializations of templates outside the specifications.

The scope of this PR is to enhance the specification and better describe the annotations we have at Jakarta NoSQL. Furthermore, we focus on the Template interface.

If the team agrees with this enhancement, we can go further to the TCK.

### CODE REVIEW

File: api/src/main/java/jakarta/nosql/Template.java [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+     * about the state of the entity instances that are supplied in the parameter.
+     * The position of entities within the {@code Iterable} return value must correspond to the
+     * position of entities in the parameter based on the unique identifier of the entity.</p>
+     *
+     * @param entities entities to insert.
+     * @param <T>      the entity type
+     * @param ttl      time to live
+     * @return an iterable containing the inserted entities, which may or may not be different instances depending
+     * on whether the insert caused values to be generated or automatically incremented.
+     * @throws NullPointerException if the iterable is null or any element is null.
```
Comment (njr-11) [inline api/src/main/java/jakarta/nosql/Template.java:186]: In your other insert method with time-to-live, you listed UnsupportedOpeationException. Do you want that here as well?
```suggestion
     * @throws NullPointerException if the iterable is null or any element is null.
     * @throws UnsupportedOperationException if the database does not provide time-to-live for insert operations.

```

File: api/src/main/java/jakarta/nosql/Template.java [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+     * that is supplied as a parameter. This method makes no guarantees about the state of the
+     * instance that is supplied as a parameter.</p>
+     *
+     * @param entity the entity to insert. Must not be {@code null}.
+     * @param ttl    time to live
+     * @param <T>    the entity type
+     * @return the inserted entity, which may or may not be a different instance depending on whether the insert caused
+     * values to be generated or automatically incremented.
+     * @throws NullPointerException          if the entity is null.
+     * @throws UnsupportedOperationException when the database does not provide TTL
```
Comment (njr-11) [inline api/src/main/java/jakarta/nosql/Template.java:139]: What should the user expect to happen if the time-to-live is exceeded? This method doesn't specify.

File: api/src/main/java/jakarta/nosql/Template.java [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    <T> Iterable<T> insert(Iterable<T> entities, Duration ttl);
+
+    /**
+     * Modifies an entity that already exists in the database.
+     *
+     * <p>For an update to be made, a matching entity with the same unique identifier
+     * must be present in the database. In databases that use an append model to write data or
+     * follow the BASE model, this method behaves the same as the {@link #insert} method.</p>
+     *
+     * <p>If the entity is versioned (for example, with {@code jakarta.persistence.Version} or by
```
Comment (njr-11) [inline api/src/main/java/jakarta/nosql/Template.java:197]: Remove reference to jakarta.persistence class. Does NoSQL have an equivalent?
```suggestion
     * <p>If the entity is versioned (for example, with an annotation or by
```

File: spec/src/main/asciidoc/chapters/api/annotations.adoc [hunk 4]
```diff
@@ -75,43 +125,182 @@ public class Address {
...
-This annotation defines which fields that belong to an Entity will be persisted. The field name specifies the column name by default.
+Entities can also be associated using collection types like `Iterable`, such as `List` or `Set`. Let's consider an example where an `Owner` entity is associated with multiple `Car` entities:
 
 [source,java]
 ----
+@Entity
+public class Owner {
+    @Id
+    private String name;
+   @Column
```
Comment (njr-11) [inline spec/src/main/asciidoc/chapters/api/annotations.adoc:154]:
```suggestion
    @Column
```

File: spec/src/main/asciidoc/chapters/api/annotations.adoc [hunk 5]
```diff
@@ -131,40 +320,303 @@ public class Person {
...
+    @Column
+    private Job job;
+
+    @Column("money")
+    @Convert(MoneyConverter.class)
+    private MonetaryAmount salary;
+}
+----
+
+In this example, the `Employee' entity's' salary' field is annotated with `@Convert`, specifying the `MoneyConverter` class as the converter. The `MoneyConverter` class implements the `AttributeConverter` interface to convert `MonetaryAmount` objects to and from their database representation.
```
Comment (njr-11) [inline spec/src/main/asciidoc/chapters/api/annotations.adoc:372]:
```suggestion
In this example, the `Employee` entity's salary field is annotated with `@Convert`, specifying the `MoneyConverter` class as the converter. The `MoneyConverter` class implements the `AttributeConverter` interface to convert `MonetaryAmount` objects to and from their database representation.
```

File: spec/src/main/asciidoc/chapters/api/annotations.adoc [hunk 6]
```diff
@@ -131,40 +320,303 @@ public class Person {
...
     @Column
     private String name;
+}
+
+@MappedSuperclass
+public class Animal {
 
     @Column
-    private List<String> phones;
+    private String race;
```
Comment (njr-11) [inline spec/src/main/asciidoc/chapters/api/annotations.adoc:450]: Race is not the correct term. This should probably be breed or species.
```suggestion
    private String breed;
```

File: spec/src/main/asciidoc/chapters/api/annotations.adoc [hunk 7]
```diff
@@ -131,40 +320,303 @@ public class Person {
...
+
+    @Column
+    private Integer age;
 }
 ----
 
-And, like `@Column`, if the ID’s name requires customization, it just set the single attribute of the annotation to specify the desired name:
+In this example, the `Animal` class serves as a mapped superclass where common attributes like `race` and `age` are defined. Subclasses, such as `Dog`, can then inherit these mappings, enabling a consistent data model across entities while accommodating the flexibility of NoSQL database structures.
```
Comment (njr-11) [inline spec/src/main/asciidoc/chapters/api/annotations.adoc:457]:
```suggestion
In this example, the `Animal` class serves as a mapped superclass where common attributes like `breed` and `age` are defined. Subclasses, such as `Dog`, can then inherit these mappings, enabling a consistent data model across entities while accommodating the flexibility of NoSQL database structures.
```

File: spec/src/main/asciidoc/chapters/api/annotations.adoc [hunk 8]
```diff
@@ -131,40 +320,303 @@ public class Person {
...
+
+Here's a JSON sample demonstrating the usage of `Animal` in a subclass `Dog`:
+
+[source,json]
+----
+{
+   "race": "Golden Retriever",
```
Comment (njr-11) [inline spec/src/main/asciidoc/chapters/api/annotations.adoc:464]:
```suggestion
   "breed": "Golden Retriever",
```

File: spec/src/main/asciidoc/chapters/api/annotations.adoc [hunk 9]
```diff
@@ -131,40 +320,303 @@ public class Person {
...
+   "age": 3,
+   "name": "Buddy"
+}
+----
+
+In this JSON representation, the attributes `race` and `age` from the `Animal` superclass are inherited by the `Dog` entity, showcasing the application of mapped superclass mappings to its subclasses.
```
Comment (njr-11) [inline spec/src/main/asciidoc/chapters/api/annotations.adoc:470]:
```suggestion
In this JSON representation, the attributes `breed` and `age` from the `Animal` superclass are inherited by the `Dog` entity, showcasing the application of mapped superclass mappings to its subclasses.
```

File: spec/src/main/asciidoc/chapters/api/entity.adoc [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+- the embeddable class lacks its own persistent identity, and
+- the state of an instance of the embeddable class can only be stored in the database when the instance is referenced directly or indirectly by a "parent" entity class instance.
+
+An _embedded field_ is a field whose type is an embeddable class.
+
+Like entities, embeddable classes may have basic fields, embeddable fields, and association fields, but, unlike entities, they do not have identifier fields.
+
+Like entities, a programming model for entity classes might support mutable embeddable classes, immutable embeddable classes, or both.
+
+A programming model for entity classes might define an annotation that identifies a user-written class as an embeddable class. For example, Jakarta Persistence defines the annotation `jakarta.persistence.Embeddabe`. Alternatively, the programming model might define an annotation that identifies a field as an embedded field. For example, Jakarta Persistence defines the annotation `jakarta.persistence.Embedded`.
```
Comment (njr-11) [inline spec/src/main/asciidoc/chapters/api/entity.adoc:133]:
```suggestion
Jakarta NoSQL defines an annotation that identifies a user-written class as an embeddable class: `jakarta.nosql.Embeddable`.
```

**Review by njr-11 — CHANGES_REQUESTED** [review]
Nice alignment of entity-related annotations with Jakarta Persistence.
I added a number of review comments which are mostly minor details/grammar.

File: api/src/main/java/jakarta/nosql/Id.java [hunk 11]
```diff
@@ -22,41 +22,57 @@
...
-     *
-     *     @Id("userId")
-     *     private String userName;
-     *
      * }
-     * }</pre>
-     *
      * @return the entity ID name
      */
     String value() default "_id";
```
[the remark this answers is not in the document]
Comment (otaviojava) [reply api/src/main/java/jakarta/nosql/Id.java:78]: I want to make it as close to Jakarta Persistence as possible.
Thus, I can remove and allow the annotation by column.

**Comment by dearrudam** [comment]
@otaviojava it's nice to see annotations like `@DiscriminatorColumn` and `@DiscriminatorValue`. Both will help developers implement the single table inheritance strategy, which strategy that is commonly implemented when they're working with NoSQL databases like AWS Dynamo DB.

File: api/src/main/java/jakarta/nosql/Template.java (hunk 2 above)
Comment (njr-11) [reply api/src/main/java/jakarta/nosql/Template.java:144]: Excellent, now the user will know what to expect of it. I'll mark this conversation resolved.

**Comment by otaviojava** [comment]
> FYI - just in case they were missed in the large amount of comments, there are 2 very minor suggested changes pending. One is a fix to a copyright year and the other is a grammar correction.

Thank you @njr-11. I fixed it all.
There is a lot of work to do here, but with this sync with JPA, we are going on the right path.

**Review by njr-11 — APPROVED** [review]

**Pull request merged**